[macOS] "Update AutoFill Information" prompt incorrectly shows up after performing AutoFill and then closing a tab

whsieh opened the pull request:
#### 97eba3acfe770439b967d2ce82b27c27d2c5fc2e
<pre>
[macOS] &quot;Update AutoFill Information&quot; prompt incorrectly shows up after performing AutoFill and then closing a tab
<a href="https://bugs.webkit.org/show_bug.cgi?id=288315">https://bugs.webkit.org/show_bug.cgi?id=288315</a>
<a href="https://rdar.apple.com/145085437">rdar://145085437</a>

Reviewed by Abrar Rahman Protyasha.

To improve web compatibility, Safari&apos;s AutoFill JavaScript creates and dispatches `TextEvent` on
text fields in order to simulate text insertion. This internally triggers typing commands, which
marks the text field as having been last modified via user edit. This, in turn, causes Safari to
(incorrectly) believe that the user had manually modified the AutoFilled text field after invoking
AutoFill, which then shows a prompt.

To fix this, we add some plumbing from `TextEvent` → `TypingCommand` → `Editor::appliedEditing` to
avoid marking the text field as having been modified through user editing, only in the case where
these text editing commands were triggered by a `TextEvent` that was created and initialized from
JS bindings.

* LayoutTests/fast/forms/textfield-lastchange-was-useredit-expected.txt:
* LayoutTests/fast/forms/textfield-lastchange-was-useredit.html:

Augment an existing layout test to cover this scenario.

* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
* Source/WebCore/dom/TextEvent.h:

Add a flag to mark a `TextEvent`, as being created from JS bindings (as opposed to created from the
engine as a result of handling key events, etc.).

* Source/WebCore/editing/Editor.cpp:
(WebCore::notifyTextFromControls):
(WebCore::Editor::appliedEditing):

If the edit command is a typing command that was triggered by an event created from bindings, pass
`wasUserEdit := false`.

(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeInsertText):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::insertText):

Add plumbing to send `TextEvent*` through `TypingCommand::insertText`, so that we can mark the
`TypingCommand` as being triggered by a text event that was created from bindings.

* Source/WebCore/editing/TypingCommand.h:

Add a flag to mark `TypingCommand` as being triggered by an event that was created from bindings.

(isType):

Make it possible to use `dynamicDowncast&lt;TypingCommand&gt;(…)` on an edit command.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::didEditInnerTextValue):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Plumb `wasUserEdit` down into the text form control, and set `m_lastChangeWasUserEdit`.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::insertUnicodeCharacter):

Canonical link: <a href="https://commits.webkit.org/290927@main">https://commits.webkit.org/290927@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f3ef0c8705aed32de466d317d8b50e06e927099

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91480 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/535 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96449 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42168 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93530 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11389 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19401 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70244 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27763 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8685 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82869 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50570 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/456 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41338 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78769 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98452 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18642 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79265 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18897 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78469 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19411 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22991 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/353 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11772 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18640 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23916 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18350 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20116 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/Editor.cpp:
```diff
@@ -1246,7 +1246,13 @@ void Editor::appliedEditing(CompositeEditCommand& command)
     VisibleSelection newSelection(command.endingSelection());
 
-    notifyTextFromControls(composition->startingRootEditableElement(), composition->endingRootEditableElement());
+    bool shouldNotifyFormControls = [&] {
```
optional, consider capturing `command` only rather than all the `Editor` state

whsieh replied:
Will do — thanks for the review!

whsieh replied:
I also just realized — this patch probably needs a slight adjustment to call `subtreeHasChanged()` in `didEditInnerTextValue` here, otherwise things like form validation won't trigger:

```
+void HTMLTextFormControlElement::didEditInnerTextValue(bool wasUserEdit)
 {
     if (!renderer() || !isTextField())
         return;

     LOG(Editing, "HTMLTextFormControlElement %p didEditInnerTextValue", this);

-    m_lastChangeWasUserEdit = true;
+    m_lastChangeWasUserEdit = wasUserEdit;
     subtreeHasChanged();
 }
```

I'm adjusting this patch so that it passes a `wasUserEdit` flag down to `didEditInnerTextValue`, to set the value there.

aprotyas replied:
Looks good!

whsieh committed:
[macOS] "Update AutoFill Information" prompt incorrectly shows up after performing AutoFill and then closing a tab
https://bugs.webkit.org/show_bug.cgi?id=288315
rdar://145085437

Reviewed by Abrar Rahman Protyasha.

To improve web compatibility, Safari's AutoFill JavaScript creates and dispatches `TextEvent` on
text fields in order to simulate text insertion. This internally triggers typing commands, which
marks the text field as having been last modified via user edit. This, in turn, causes Safari to
(incorrectly) believe that the user had manually modified the AutoFilled text field after invoking
AutoFill, which then shows a prompt.

To fix this, we add some plumbing from `TextEvent` → `TypingCommand` → `Editor::appliedEditing` to
avoid marking the text field as having been modified through user editing, only in the case where
these text editing commands were triggered by a `TextEvent` that was created and initialized from
JS bindings.

* LayoutTests/fast/forms/textfield-lastchange-was-useredit-expected.txt:
* LayoutTests/fast/forms/textfield-lastchange-was-useredit.html:

Augment an existing layout test to cover this scenario.

* Source/WebCore/dom/TextEvent.cpp:
(WebCore::TextEvent::TextEvent):
* Source/WebCore/dom/TextEvent.h:

Add a flag to mark a `TextEvent`, as being created from JS bindings (as opposed to created from the
engine as a result of handling key events, etc.).

* Source/WebCore/editing/Editor.cpp:
(WebCore::notifyTextFromControls):
(WebCore::Editor::appliedEditing):

If the edit command is a typing command that was triggered by an event created from bindings, pass
`wasUserEdit := false`.

(WebCore::Editor::insertTextWithoutSendingTextEvent):
(WebCore::Editor::setComposition):
* Source/WebCore/editing/EditorCommand.cpp:
(WebCore::executeInsertText):
* Source/WebCore/editing/TypingCommand.cpp:
(WebCore::TypingCommand::insertText):

Add plumbing to send `TextEvent*` through `TypingCommand::insertText`, so that we can mark the
`TypingCommand` as being triggered by a text event that was created from bindings.

* Source/WebCore/editing/TypingCommand.h:

Add a flag to mark `TypingCommand` as being triggered by an event that was created from bindings.

(isType):

Make it possible to use `dynamicDowncast<TypingCommand>(…)` on an edit command.

* Source/WebCore/html/HTMLTextFormControlElement.cpp:
(WebCore::HTMLTextFormControlElement::didEditInnerTextValue):
* Source/WebCore/html/HTMLTextFormControlElement.h:

Plumb `wasUserEdit` down into the text form control, and set `m_lastChangeWasUserEdit`.

* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::insertUnicodeCharacter):

Canonical link: https://commits.webkit.org/290927@main
